### PR TITLE
Finish deprecating the fs argument

### PR DESCRIPTION
**Pull request by dconathan**
See #5385 for some discussion on this

The `fs=` arg was depcrecated from `Dataset.save_to_disk` and `Dataset.load_from_disk` in `2.8.0` (to be removed in `3.0.0`).  There are a few other places where the `fs=` arg was still used (functions/methods in `datasets.info` and `datasets.load`).  This PR adds a similar behavior, warnings and the `storage_options=` arg to these functions and methods.

One question: should the "deprecated" / "added" versions be `2.8.1` for the docs/warnings on these?  Right now I'm going with "fs was deprecated in 2.8.0" but "storage_options= was added in 2.8.1" where appropriate.

@mariosasko

### CODE REVIEW

File: src/datasets/info.py [hunk 1]
```diff
@@ -302,7 +337,17 @@ def from_directory(cls, dataset_info_dir: str, fs=None) -> "DatasetInfo":
         >>> ds_info = DatasetInfo.from_directory("/path/to/directory/")
         ```
         """
-        fs = fs or LocalFileSystem()
+        if fs != "deprecated":
+            warnings.warn(
+                "'fs' was is deprecated in favor of 'storage_options' in version 2.8.0 and will be removed in 3.0.0.\n"
```
Comment (albertvillanova) [inline src/datasets/info.py:342]:
```suggestion
                "'fs' was deprecated in favor of 'storage_options' in version 2.9.0 and will be removed in 3.0.0.\n"
```

File: src/datasets/load.py [hunk 2]
```diff
@@ -1787,13 +1789,26 @@ def load_from_disk(dataset_path: str, fs=None, keep_in_memory: Optional[bool] =
             Path (e.g. `"dataset/train"`) or remote URI (e.g.
             `"s3://my-bucket/dataset/train"`) of the [`Dataset`] or [`DatasetDict`] directory where the dataset will be
             loaded from.
-        fs (`~filesystems.S3FileSystem` or `fsspec.spec.AbstractFileSystem`, *optional*, defaults to `None`):
+        fs (`~filesystems.S3FileSystem` or `fsspec.spec.AbstractFileSystem`, *optional*):
             Instance of the remote filesystem used to download the files from.
+
+            <Deprecated version="2.8.0">
```
Comment (albertvillanova) [inline src/datasets/load.py:1795]:
```suggestion
            <Deprecated version="2.9.0">
```

File: src/datasets/load.py [hunk 3]
```diff
@@ -1787,13 +1789,26 @@ def load_from_disk(dataset_path: str, fs=None, keep_in_memory: Optional[bool] =
...
+
+            `fs` was deprecated in version 2.8.0 and will be removed in 3.0.0.
```
Comment (albertvillanova) [inline src/datasets/load.py:1797]:
```suggestion
            `fs` was deprecated in version 2.9.0 and will be removed in 3.0.0.
```

File: src/datasets/load.py [hunk 4]
```diff
@@ -1787,13 +1789,26 @@ def load_from_disk(dataset_path: str, fs=None, keep_in_memory: Optional[bool] =
...
+            Please use `storage_options` instead, e.g. `storage_options=fs.storage_options`
```
Comment (albertvillanova) [inline src/datasets/load.py:1798]:
```suggestion
            Please use `storage_options` instead, e.g. `storage_options=fs.storage_options`.
```

File: src/datasets/load.py [hunk 5]
```diff
@@ -1787,13 +1789,26 @@ def load_from_disk(dataset_path: str, fs=None, keep_in_memory: Optional[bool] =
...
+
         keep_in_memory (`bool`, defaults to `None`):
             Whether to copy the dataset in-memory. If `None`, the dataset
             will not be copied in-memory unless explicitly enabled by setting `datasets.config.IN_MEMORY_MAX_SIZE` to
             nonzero. See more details in the [improve performance](./cache#improve-performance) section.
 
+        storage_options (`dict`, *optional*):
+            Key/value pairs to be passed on to the file-system backend, if any.
+
+            <Added version="2.8.1"/>
```
Comment (albertvillanova) [inline src/datasets/load.py:1810]:
```suggestion
            <Added version="2.9.0"/>
```

File: src/datasets/load.py [hunk 6]
```diff
@@ -1806,6 +1821,17 @@ def load_from_disk(dataset_path: str, fs=None, keep_in_memory: Optional[bool] =
     >>> ds = load_from_disk('path/to/dataset/directory')
     ```
     """
+    if fs != "deprecated":
+        warnings.warn(
+            "'fs' was is deprecated in favor of 'storage_options' in version 2.8.0 and will be removed in 3.0.0.\n"
```
Comment (albertvillanova) [inline src/datasets/load.py:1826]:
```suggestion
            "'fs' was deprecated in favor of 'storage_options' in version 2.9.0 and will be removed in 3.0.0.\n"
```

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks for the deprecation. Some minor suggested fixes below...

Also note that the corresponding tests should be updated as well.

**Comment by dconathan** [comment]
> Thanks for the deprecation. Some minor suggested fixes below...
> 
> Also note that the corresponding tests should be updated as well.

Thanks for the suggestions/typo fixes.  I updated the failing test - passing locally now

**Comment by lhoestq** [comment]
Nice thanks !

I believe you also need to update `_load_info` and `_save_info` in `builder.py` - they're still passing `fs=self._fs` instead of `storage_options=self._fs.storage_options`

This should remove the remaining warnings in the CI such as 

```python
tests/test_builder.py::test_builder_with_filesystem_download_and_prepare_reload
tests/test_load.py::test_load_dataset_local[False]
tests/test_load.py::test_load_dataset_local[True]
tests/test_load.py::test_load_dataset_zip_csv[csv_path-False]
tests/test_load.py::test_load_dataset_then_move_then_reload
  /opt/hostedtoolcache/Python/3.7.15/x64/lib/python3.7/site-packages/datasets/info.py:344: FutureWarning: 'fs' was deprecated in favor of 'storage_options' in version 2.9.0 and will be removed in 3.0.0.
  You can remove this warning by passing 'storage_options=fs.storage_options' instead.
```

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks again for all the work, @dconathan.

I agree with @lhoestq that we should better address all remaining `fs` deprecation warnings with this PR.

For example, there are still some deprecation warnings when calling `Dataset.load_from_disk` with `fs`. See: https://github.com/huggingface/datasets/blob/232a43943e87dfedcc328a9a3d3b4d89ea5c6627/src/datasets/load.py#L1819

or `DatasetDict.load_from_disk` with `fs`. See: https://github.com/huggingface/datasets/blob/232a43943e87dfedcc328a9a3d3b4d89ea5c6627/src/datasets/load.py#L1821

These docstrings should also be updated: 
https://github.com/huggingface/datasets/blob/232a43943e87dfedcc328a9a3d3b4d89ea5c6627/src/datasets/filesystems/s3filesystem.py#L98
https://github.com/huggingface/datasets/blob/232a43943e87dfedcc328a9a3d3b4d89ea5c6627/src/datasets/filesystems/s3filesystem.py#L111

**Comment by dconathan** [comment]
re: docstring, I assume passing in `storage_options=s3.storage_options` is correct/necessary to pass the secrets?

**Comment by dconathan** [comment]
what about 
https://github.com/huggingface/datasets/blob/5b793dd8c43bf6e85f165238becb3c64f6cd3ed0/src/datasets/filesystems/__init__.py#L43-L54

**Review by albertvillanova — APPROVED** [review]
Thanks again for all your work on this PR, @dconathan.

I think the function `is_remote_filesystem` should be kept as it is.

We are going to re-run the CI. Once all green, we can merge.

**Pull request merged**